src/commands/farm.rs

```diff
@@ -239,30 +239,28 @@ async fn subscribe_to_plotting_progress(
 }
 
 async fn subscribe_to_solutions(
-    summary: SummaryFilePointer,
+    mut summary: Summary,
     node: Arc<Node>,
     is_initial_progress_finished: Arc<AtomicBool>,
     reward_address: PublicKey,
 ) -> Result<()> {
+    // necessary for spacing
     println!();
-    let mut new_blocks = node
-        .subscribe_new_blocks()
-        .await
-        .into_eyre()
-        .context("couldn't subscribe to new blocks")?;
 
-    while let Some(new_block) = new_blocks.next().await {
-        let mut summary_update_values: SummaryUpdateFields = Default::default();
+    let SummaryInner { last_block_num, .. } =
+        summary.parse_summary().await.context("couldn't acquire the lock")?;
 
-        let events = node
-            .get_events(Some(new_block.hash))
+    for block_num in last_block_num as u32
+        ..node.get_info().await.into_eyre().context("couldn't get block from node")?.best_block.1
+    {
+        let mut summary_update_values: SummaryUpdateFields = Default::default();
+        let block_hash = node.block_hash(block_num).into_eyre()?.expect("Block hash always exists");
+        for event in node
+            .get_events(Some(block_hash))
             .await
             .into_eyre()
-            .context("couldn't get events from node")?;
-
-        for event in events {
-            // subscription is active when plotting is started, only print out rewards after
-            // plotting finishes to not corrup the progress bars
+            .context("couldn't get event from node")?
+        {
```

Comment:
I would parallelize it using [async_stream](https://docs.rs/async-stream/latest/async_stream/):
```rust
    fn missed_blocks_stream(
        node: std::sync::Arc<Node>,
        mut last_block_num: subspace_sdk::node::BlockNumber,
    ) -> impl Stream<Item = anyhow::Result<subspace_sdk::node::BlockNumber>> {
        async_stream::try_stream! {
            loop {
                let last_retrieved_num = node.get_info().await?.finalized_block.1;

                if last_block_num == last_retrieved_num {
                    break;
                }

                while last_block_num < last_retrieved_num {
                    yield last_block_num;
                    last_block_num += 1;
                }
            }
        }
    }

    let concurency_parameter = todo!();

    let node = std::sync::Arc::new(todo!());
    let s = missed_blocks_stream(std::sync::Arc::clone(&node), todo!());

    futures::pin_mut!(s);

    let missed_rewards: u128 = s
        .map(|block_number| {
            Ok(node.get_events(Some(
                node.block_hash(block_number?)?
                    .expect("Should be always present if no pruning enabled"),
            )))
        })
        .try_buffer_unordered(concurency_parameter)
        .map_ok(|events| futures::stream::iter(events).map(anyhow::Result::Ok))
        .try_flatten()
        .map_ok(|event| match event {
            Event::Rewards(
                RewardsEvent::VoteReward { voter: author, reward }
                | RewardsEvent::BlockReward { block_author: author, reward },
            ) if author == reward_address.into() => reward,
            _ => 0,
        })
        .try_fold(0u128, |acc, rewards| {
            futures::future::ready::<anyhow::Result<_>>(Ok(acc + rewards))
        })
        .await?;
```

Reply:
Sorry, missed part with farmed block count here, but I think you get an idea.

Reply:
You can update summary somewhere in between here